File: ui/src/lib/protocol.ts
/**
 * Vision-RCP Protocol — TypeScript type definitions.
 * Mirrors daemon/src/protocol.py exactly.
 */

export type MessageType = 'command' | 'response' | 'stream' | 'error' | 'heartbeat';

export type CommandType =
  | 'process.spawn'
  | 'process.kill'
  | 'process.restart'
  | 'process.list'
  | 'process.status'
  | 'process.logs'
  | 'graph.start'
  | 'graph.stop'
  | 'graph.status'
  | 'auth.login'
  | 'auth.refresh'
  | 'auth.logout'
  | 'system.info'
  | 'system.ping'
  | 'audit.query'
  | 'agent.start'
  | 'agent.send'
  | 'agent.interrupt'
  | 'agent.stop'
  | 'agent.approve'
  | 'session.info';

export type ProcessState =
  | 'pending'
  | 'starting'
  | 'running'
  | 'stopping'
  | 'stopped'
  | 'failed'
  | 'restarting';

export type StreamType = 'stdout' | 'stderr' | 'system';

export interface RCPError {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

export interface Envelope {
  id: string;
  type: MessageType;
  ts: string;
  token?: string;
  command?: CommandType;
  ref?: string;
  payload?: Record<string, unknown>;
  error?: RCPError;
}

export interface ProcessInfo {
  name: string;
  cmd: string;
  args: string[];
  cwd?: string;
  depends_on: string[];
  auto_restart: boolean;
  group?: string;
  pid?: number;
  state: ProcessState;
  started_at?: number;
  restart_count: number;
  exit_code?: number;
  uptime: number;
  cpu_percent: number;
  memory_rss: number;
  memory_vms: number;
}

export interface SystemInfo {
  os: string;
  platform: string;
  hostname: string;
  cpu_count: number;
  cpu_percent: number;
  memory_total: number;
  memory_used: number;
  memory_percent: number;
  disk_total: number;
  disk_used: number;
  disk_percent: number;
  uptime: number;
  daemon_port: number;
}

export interface LogLine {
  pid: number;
  name: string;
  stream: StreamType;
  data: string;
  ts: number;
}

export interface GraphNode {
  id: string;
  state: ProcessState;
  pid?: number;
}

export interface GraphEdge {
  from: string;
  to: string;
}

export interface GraphStatus {
  name: string;
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export interface AgentEvent {
  type: 'AGENT_MESSAGE' | 'APPROVAL_REQUEST' | 'AGENT_STATUS';
  content: string;
  timestamp: number;
  source?: 'user' | 'agent' | 'system';
}

export interface SessionInfo {
  sessionId: string;
  deviceName: string;
  workspace: string;
  agentName: string | null;
  agentStatus: string;
  availableAgents: string[];
  mode: 'local' | 'relay';
  connectedAt: string;
}

/**
 * Create a command envelope.
 */
export function createCommand(
  command: CommandType,
  payload: Record<string, unknown> = {},
  token?: string,
): Envelope {
  return {
    id: crypto.randomUUID(),
    type: 'command',
    ts: new Date().toISOString(),
    command,
    payload,
    token,
  };
}

File: daemon/src/protocol.py
"""Vision-RCP Protocol — Message definitions and validation."""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class MessageType(str, Enum):
    COMMAND = "command"
    RESPONSE = "response"
    STREAM = "stream"
    ERROR = "error"
    HEARTBEAT = "heartbeat"
    HANDSHAKE = "handshake"


class CommandType(str, Enum):
    PROCESS_SPAWN = "process.spawn"
    PROCESS_KILL = "process.kill"
    PROCESS_RESTART = "process.restart"
    PROCESS_LIST = "process.list"
    PROCESS_STATUS = "process.status"
    PROCESS_LOGS = "process.logs"
    GRAPH_START = "graph.start"
    GRAPH_STOP = "graph.stop"
    GRAPH_STATUS = "graph.status"
    AUTH_LOGIN = "auth.login"
    AUTH_REFRESH = "auth.refresh"
    AUTH_LOGOUT = "auth.logout"
    SYSTEM_INFO = "system.info"
    SYSTEM_PING = "system.ping"
    
    # Audit Logs
    AUDIT_QUERY = "audit.query"

    # Agent Commands
    AGENT_START = "agent.start"
    AGENT_SEND = "agent.send"
    AGENT_INTERRUPT = "agent.interrupt"
    AGENT_STOP = "agent.stop"
    AGENT_APPROVE = "agent.approve"
    SESSION_INFO = "session.info"
    SESSION_LIST_CLIENTS = "session.clients"


class ProcessState(str, Enum):
    PENDING = "pending"
    STARTING = "starting"
    RUNNING = "running"
    STOPPING = "stopping"
    STOPPED = "stopped"
    FAILED = "failed"
    RESTARTING = "restarting"


class StreamType(str, Enum):
    STDOUT = "stdout"
    STDERR = "stderr"
    SYSTEM = "system"


class RCPError(BaseModel):
    code: str
    message: str
    details: Optional[dict[str, Any]] = None


class Envelope(BaseModel):
    """RCP protocol message envelope. Every message conforms to this shape."""
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    type: MessageType
    ts: str = Field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    token: Optional[str] = None
    command: Optional[CommandType] = None
    ref: Optional[str] = None
    payload: Optional[dict[str, Any]] = None
    error: Optional[RCPError] = None

    @classmethod
    def cmd(cls, command: CommandType, payload: dict[str, Any] | None = None,
            token: str | None = None) -> "Envelope":
        return cls(type=MessageType.COMMAND, command=command,
                   payload=payload or {}, token=token)

    @classmethod
    def ok(cls, ref: str, payload: dict[str, Any] | None = None) -> "Envelope":
        return cls(type=MessageType.RESPONSE, ref=ref, payload=payload or {})

    @classmethod
    def err(cls, ref: str, code: str, message: str,
            details: dict[str, Any] | None = None) -> "Envelope":
        return cls(type=MessageType.ERROR, ref=ref,
                   error=RCPError(code=code, message=message, details=details))

    @classmethod
    def stream_msg(cls, ref: str, payload: dict[str, Any]) -> "Envelope":
        return cls(type=MessageType.STREAM, ref=ref, payload=payload)

    @classmethod
    def heartbeat(cls) -> "Envelope":
        return cls(type=MessageType.HEARTBEAT)

    def to_json(self) -> str:
        return self.model_dump_json(exclude_none=True)

    @classmethod
    def from_json(cls, data: str) -> "Envelope":
        return cls.model_validate_json(data)
